microscope/monitor/runner.py

```diff
@@ -0,0 +1,300 @@
+        self.close_queue = Queue()
+
+    def run(self, monitor_args: MonitorArgs, nodes: List[str],
+            cmd_override: str):
+
+        api = core_v1_api.CoreV1Api()
+
+        try:
+            pods = api.list_namespaced_pod(self.namespace,
+                                           label_selector='k8s-app=cilium')
+        except ApiException as e:
+            print('could not list Cilium pods: %s\n' % e)
+            sys.exit(1)
+
+        if nodes:
+            names = [(pod.metadata.name, pod.spec.node_name)
+                     for pod in pods.items
+                     if
+                     pod.metadata.name in nodes
+                     or
+                     pod.spec.node_name in nodes]
+        else:
+            names = [(pod.metadata.name, pod.spec.node_name)
+                     for pod in pods.items]
+
+        if not names:
+            raise ValueError('No Cilium nodes in cluster match provided names'
+                             ', or Cilium is not deployed')
+
+        endpoints = self.retrieve_endpoint_data()
+
+        if cmd_override:
+            cmd = cmd_override.split(" ")
+        else:
+            cmd = self.get_monitor_command(monitor_args,
+                                           [name[0] for name in names],
+                                           endpoints)
+        mode = ""
+
+        if monitor_args.raw:
+            mode = "raw"
+
+        if monitor_args.verbose or cmd_override:
+            mode = "verbose"
+
+        identities = self.retrieve_identities(endpoints)
+        endpoint_info = self.retrieve_endpoint_info(endpoints)
+        self.monitors = [
+            Monitor(name[0], name[1], self.namespace, self.data_queue,
+                    self.close_queue, api, cmd, mode, identities,
+                    endpoint_info)
+            for name in names]
+
+        for m in self.monitors:
+            m.process.start()
+
+    def retrieve_identities(self, endpoint_data: Dict) -> Dict:
+        labels = {id["id"]: id["labels"] for id in
+                  [e["status"]["status"]["identity"]
+                   for e in endpoint_data["items"]]}
+
+        labels[0] = ["reserved:unknown"]
+        labels[1] = ["reserved:host"]
+        labels[2] = ["reserved:world"]
+        labels[3] = ["reserved:cluster"]
+        labels[4] = ["reserved:health"]
+
+        return labels
+
+    def retrieve_endpoint_info(self, endpoint_data: Dict) -> Dict:
+        return {x["status"]["id"]:
+                {
+                    "name": x["metadata"]["name"],
+                    "namespace": x["metadata"]["namespace"],
+                    "networking": x["status"]["status"]["networking"]
+                }
+                for x in endpoint_data["items"]}
+
+    def get_monitor_command(self, args: MonitorArgs, names: List[str],
+                            endpoint_raw_data: Dict) -> List[str]:
+        endpoints = [[cep['status'] for cep in endpoint_raw_data['items']]]
```

Comment:
Any reason we need this nested list in list? Sorry if I'm missing something in the code below. But I see that `retrieve_endpoint_ids` method has just one forloop which is going to iterate once over `endpoints`.